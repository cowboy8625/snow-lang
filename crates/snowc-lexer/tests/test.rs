use snowc_lexer::{Scanner, Token};

pub fn snapshot_lexing(input: &str) -> String {
    let lexer = Scanner::new(input);
    let tokens = lexer.collect::<Vec<Token>>();
    let mut tokens = std::collections::VecDeque::from(tokens);
    let mut output = String::new();
    for (row, line) in input.lines().enumerate() {
        output += line;
        output += "\n";
        while let Some(tok) = tokens.pop_front() {
            if tok.span().row_end != row {
                tokens.push_front(tok);
                break;
            }
            output += &" ".repeat(tok.span().col_start);
            output += &"^".repeat(tok.span().len()); // tok.span().col_end - tok.span().col_start);
            output += &format!(" {tok:?}");
            output += "\n"
        }
    }
    output
}

macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            let contents = include_str!($path);
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(snapshot_lexing(contents));
            });
        }
    };
}

snapshot!(hello_world, "./../../../samples/hello_world.snow");
