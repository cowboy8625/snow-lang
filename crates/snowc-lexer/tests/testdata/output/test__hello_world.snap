---
source: crates/snowc-lexer/tests/test.rs
expression: snapshot_lexing(contents)
---
main : IO = print "Winter is Coming!"
^^^^ Ident(Ident { lexme: "main", pos: Start, span: (0,0)->(4,0) })
     ^ Ctrl(Ctrl { lexme: ":", pos: Middle, span: (5,0)->(6,0) })
       ^^ Ident(Ident { lexme: "IO", pos: Middle, span: (7,0)->(9,0) })
          ^ Ctrl(Ctrl { lexme: "=", pos: Middle, span: (10,0)->(11,0) })
            ^^^^^ Ident(Ident { lexme: "print", pos: Middle, span: (12,0)->(17,0) })
                  ^^^^^^^^^^^^^^^^^^^ Str(Str { lexme: "Winter is Coming!", pos: End, span: (18,0)->(37,0) })

